### PR TITLE
Possible Exercises

### PR DESCRIPTION
**Pull request by shhavel**
Possible exercises (without solutions).

### CODE REVIEW

File: Exercises/1-closure.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+const coffee = (volume, strength) =>
+  (`Coffee volume: ${volume}ml, strength: ${strength}`);
```
Comment (tshemsedinov) [inline Exercises/1-closure.js]:
```suggestion
  `Coffee volume: ${volume}ml, strength: ${strength}`;
```

File: Exercises/1-closure.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+// Set names for anonymous functions (for tests).
+if (typeof espresso === 'function') {
+  Object.defineProperty(espresso, 'name',
+    { value: 'espresso', configurable: true });
+}
+if (typeof americano === 'function') {
+  Object.defineProperty(americano, 'name',
+    { value: 'americano', configurable: true });
+}
```
Comment (tshemsedinov) [inline Exercises/1-closure.js]: For function `const espresso = () => {};` we can read name: `espresso.name` and it returns `espresso`.
In new version of `hpw` (exactly `0.1.11`) we don't need function names at all.

File: Exercises/1-closure.test [hunk 3]
```diff
@@ -0,0 +1,16 @@
+([{
```
Comment (tshemsedinov) [inline Exercises/1-closure.test]:
```suggestion
[{
```

File: Exercises/1-closure.test [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  ]
+},
+{
+  name: 'americano',
+  length: [30, 70],
+  cases: [
+    ['medium', 'Coffee volume: 150ml, strength: medium'],
+    ['strong', 'Coffee volume: 150ml, strength: strong'],
+  ]
+}])
```
Comment (tshemsedinov) [inline Exercises/1-closure.test]:
```suggestion
}]
```

File: Exercises/2-lambda.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+const tagged = (pref, str) => `[${pref}] ${str}`;
+
+// Define function tagDate that prepends current date to the string.
+// E.g. tagDate('My String') === '[2019-11-14] My String'
+// Use function tagged to implement tagDate.
+
+const tagDate = null;
+
+module.exports = { tagDate };
```
Comment (tshemsedinov) [inline Exercises/2-lambda.js]: :+1:

File: Exercises/2-lambda.test [hunk 6]
```diff
@@ -0,0 +1,21 @@
+({
+  name: 'tagDate',
+  length: [150, 400],
+  test: tagDate => {
+    {
+      const today = new Date();
+      const dd = String(today.getDate()).padStart(2, '0');
+      const mm = String(today.getMonth() + 1).padStart(2, '0');
+      const yyyy = today.getFullYear();
+      const expected = `[${yyyy}-${mm}-${dd}] Test`;
```
Comment (tshemsedinov) [inline Exercises/2-lambda.test]:
```suggestion
      const date = new Date().toISOString().substring(0, 10);
      const expected = `[${date}] Test`;
```

File: Exercises/3-bind.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+// Set names for anonymous functions (for tests).
+if (typeof average === 'function') {
+  Object.defineProperty(average, 'name',
+    { value: 'average', configurable: true });
+}
+if (typeof rootMeanSquare === 'function') {
+  Object.defineProperty(rootMeanSquare, 'name',
+    { value: 'rootMeanSquare', configurable: true });
+}
```
Comment (tshemsedinov) [inline Exercises/3-bind.js]: We don't need this hack, see above

File: Exercises/4-curry.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+// Check 4 digit pin.
+const checkPin = (...code) =>
+  (code.join('') === '4967');
```
Comment (tshemsedinov) [inline Exercises/4-curry.js]:
```suggestion
  code.join('') === '4967';
```

File: Exercises/4-curry.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+const curry = (length, fn) => (...args) => null;
+
+// Allows to enter pin code by one character,
+// Usage: press('3')('4')('5')('6')
+const press = curry(4, checkPin);
+
+// Set name for anonymous function (for tests).
+if (typeof press === 'function') {
+  Object.defineProperty(press, 'name', { value: 'press', configurable: true });
+}
```
Comment (tshemsedinov) [inline Exercises/4-curry.js]: Ditto

File: Exercises/5-chaining.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+// Check 4 digit pin.
+const checkPin = (ch1, ch2, ch3, ch4) =>
+  ([ch1, ch2, ch3, ch4].join('') === '4967');
```
Comment (tshemsedinov) [inline Exercises/5-chaining.js]:
```suggestion
  [ch1, ch2, ch3, ch4].join('') === '4967';
```

**Comment by shhavel** [comment]
Thanks for review. Fixed PR comments.

File: Exercises/4-curry.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+// Check 4 digit pin.
+const checkPin = (...code) =>
+  code.join('') === '4967';
```
Comment (tshemsedinov) [inline Exercises/4-curry.js]: Single-line function

File: Exercises/5-chaining.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+// Check 4 digit pin.
+const checkPin = (ch1, ch2, ch3, ch4) =>
```
Comment (tshemsedinov) [inline Exercises/5-chaining.js]:
```suggestion
const EXPECTED_PIN = '4967';
const expect = code => (...args) => args.join('') === code;
const checkPin = expect(EXPECTED_PIN);
```

File: Exercises/4-curry.js (hunk 11 above)
Comment (tshemsedinov) [reply Exercises/4-curry.js]: See example below to refactor this code: extract magic number to constant and hold it in closure.

File: Exercises/5-chaining.js (hunk 12 above)
Comment (shhavel) [reply Exercises/5-chaining.js]: I thinks the suggested change is to atomic, makes the check visually longer and therefore increases an effort required to understand the task.

How about this version?

```js
const CORRECT_PIN = '4967';
const checkPin = (...code) => code.join('') === CORRECT_PIN;
```

Comment (tshemsedinov) [reply Exercises/5-chaining.js]: Fine

**Comment by tshemsedinov** [comment]
Now it's time to reorder examples to avoid conflicts with master and then I'll land this PR @shhavel

**Comment by shhavel** [comment]
@tshemsedinov Branch was rebased with master. Examples were reordered (starting from 2). Thanks.

**Pull request merged**